DROOLS-2877: Update the top level data type, considering changes in an external/nested sub data type

karreiro opened the pull request:
See:

* [DROOLS-2877](https://issues.jboss.org/browse/DROOLS-2877): Update the top level data type, considering changes in an external/nested sub data type

* [DROOLS-2881](https://issues.jboss.org/browse/DROOLS-2881): [DMN Designer] Data type dialog - "Destroy" action in the list view

* [DROOLS-2945](https://issues.jboss.org/browse/DROOLS-2945): DMN Designer - opening data type dialog shows warning in browser console

* [DROOLS-2950](https://issues.jboss.org/browse/DROOLS-2950): [DMN Editor] Custom data-types - ItemDefinitionRecordEngine - The QName instance should lookup the real prefix from the URI

![drools-2877](https://user-images.githubusercontent.com/1079279/45698109-acb8ee00-bb3d-11e8-9f03-3830052870d7.gif)

![drools-2881](https://user-images.githubusercontent.com/1079279/45698130-b9d5dd00-bb3d-11e8-972a-4c15e17b281f.gif)

manstis commented:
Heads up... I ran (quickly) on top of DROOLS-2941 and got a NPE when dragging a Decision from the palette. Will try again tomorrow (when I'm less rushed).. but if you're still around you might want to check too.

karreiro commented:
@manstis Thank you for the heads up! :-)

..but, I could not reproduce the issue here, see:
![2018-09-18 4 52 58 pm](https://user-images.githubusercontent.com/1079279/45712874-d6850b80-bb63-11e8-95fb-b79add714b13.gif)

manstis commented:
Thanks for investigating @karreiro 

Don't try to hard to replicate. It was the first thing I encountered before anything else.

I'll try better tomorrow!

karreiro commented:
@manstis ok! Have a nice evening :-)

manstis commented:
Hi @karreiro here's observations (not reviewed the code yet):

Edit a type (click "Edit" button)
- Change data-type to something different (e.g. was "String" change to "Number")
- Click "X" to cancel change
- Text in dialog shows cancelled change
- Click "Edit" again and the data-type drop-down correctly shows the original data-type

Updated data-type of element
- Changed "tGender" from "string" to "time"
- Saved
- Examined XML
-- Actual: <semantic:typeRef>{http://www.omg.org/spec/FEEL/20140401}time</semantic:typeRef>
-- Expected: <semantic:typeRef>feel:time</semantic:typeRef>

---------------------------------------------------------------
(May be out of scope of changes made)

Select kebab menu next to a nested/child element
- Select "Remove"
- NPE

Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile")
- Change "Age" from "Number" to something else
- Save
- "tProfiles" looses its child content (i.e although showing as a "Structure" it no longer lists the content of "tProfile")
- "tProfile" Age field remains "Number"

manstis commented:
File I used is attached.

[DMCommunity Challenge - March 2017.zip](https://github.com/kiegroup/kie-wb-common/files/2396549/DMCommunity.Challenge.-.March.2017.zip)

karreiro commented:
@mantis Thank you for the feedback (and thank you for the `.dmn` file, it was really helpful).

Ok, about the issues..

> Edit a type (click "Edit" button)

Fixed, see:
![cancel-bug](https://user-images.githubusercontent.com/1079279/45866453-44ddef80-bd57-11e8-983e-7561e61e5ebf.gif)


> Updated data-type of element

Fixed, see:
![xml](https://user-images.githubusercontent.com/1079279/45866459-4ad3d080-bd57-11e8-89f6-a4e41bd2e4e6.gif)


> Select kebab menu next to a nested/child element (of top-level element); Remove it; `Cannot read property 'isTopLevel_1_g$' of undefined` (...)

Fixed, see:
![remove-tprofile](https://user-images.githubusercontent.com/1079279/45866484-60e19100-bd57-11e8-9a86-0048a28c6328.gif)


> Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile"); Save it (...)

Fixed, see:
![update-tprofiles](https://user-images.githubusercontent.com/1079279/45866490-66d77200-bd57-11e8-8daf-812661feea2c.gif)


> Edit "tProfiles" (that is of type "tProfile") and shows the child elements (defined in "tProfile"); Remove it (...)

Fixed, see:

![remove-tprofiles](https://user-images.githubusercontent.com/1079279/45866497-6dfe8000-bd57-11e8-8225-438d07570493.gif)

Thanks again for the review.

karreiro commented:
@jomarko Thank you for the review.

> Could you please add javadoc at least for new classes and public methods inside? Otherwise it is really hard to do some valuable review.
I added the Javadoc in the main classes, I hope it helps.

> The PR is really big
Yes.

> We use a lot of similar terms: handler/manager, nested/sub
I added some comments to help, but something that can help too is checking where the classes are being used, the package of the class, the pattern that each class implement or even the implementation approach.

For example, the class `DataTypeManager` looks like a _Builder_, however, it's not called a builder because besides encapsulating the object creation, we're encapsulating its interaction with Data Types when they are already created (see the `withRefreshedSubDataTypes(..)` method).

All `*Handlers` (inside of the package `org.kie.workbench.common.dmn.client.editors.types.persistence.handlers`) handle with persistence tasks. Externally, they are called only by the `ItemDefinitionRecordEngine`, which is the class responsible for recording an ItemDefinition considering the changes suffered by a Data Type).

Regarding the concepts "nested" and "sub", I see them as different things. When we're working closer to the View, the nested concept makes more sense, however, when we think how the Data Types related to each other, we can see the `subDataType` expression in the code.

> We are not consistent
> Somewhere we compare data types as a.equals(b), on other place Objects.equals(a, b)
> Somewhere we iterate lists via streams, somewhere via for cycles
I think that we could always use `Objects.equals(a, b)` instead of `a.equals(b)`. If I miss it in some place, we can fix it.
Regarding how we're handling lists in general, `stream`s and `for`s provide similar features, however sometimes one is more readable than another, there's no reason for creating strict rules about this :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandler.java:
```diff
@@ -0,0 +1,158 @@
+
+public class DataTypeHandler {
+
+    final DataTypeStore dataTypeStore;
```
This property is being used by the classes that extend `DataTypeHandler` :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionUpdateHandler.java:
```diff
@@ -0,0 +1,79 @@
+
+    QName makeQName(final DataType dataType) {
+        if (isDefault(dataType.getType())) {
+            return new QName(FEEL.getUri(), dataType.getType());
```
I think this is fixed now :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeUtils.java:
```diff
@@ -16,72 +16,45 @@
 public class DataTypeUtils {
 
-    private final DataTypeFactory dataTypeFactory;
+    private final ItemDefinitionUtils itemDefinitionUtils;
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java:
```diff
@@ -96,17 +101,57 @@ public void addSubItems(final DataType dataType,
+    void hideItemElementIfParentIsCollapsed(final HTMLElement itemElement,
+                                            final Element parent) {
+
+        final boolean isCollapsed = isCollapsed(parent.querySelector("[data-field=\"arrow-button\"]"));
```
Fixed :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListView.java:
```diff
@@ -96,17 +101,57 @@ public void addSubItems(final DataType dataType,
+
+        final Optional<Element> dataTypeRow = Optional.ofNullable(getDataTypeRow(dataType));
+
+        dataTypeRow.ifPresent(Element::remove);
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeSelectView.java:
```diff
@@ -151,17 +155,24 @@ public void setDataType(final DataType dataType) {
+
+        final String newValue = event.target.value;
+
+        if (!newValue.equals(getValue())) {
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/DataTypeStore.java:
```diff
@@ -41,7 +44,19 @@ public void clear() {
+    }
+
+    public List<DataType> all() {
+        return new ArrayList<>(dataTypes.values());
```
Fixed :-)

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/ItemDefinitionStore.java:
```diff
@@ -22,7 +22,13 @@
+ * All entries are indexed by the UUID from the correspondent Data Type.
+ */
 @ApplicationScoped
 public class ItemDefinitionStore {
```
I was comparing this class with `DataTypeStore` and noticed here we don't have `unIndex` metod. Isn't it issue as `DataTypeStore.unIndex` is used in `DataTypeDestroyHandler#destroy` and `ItemDefinitionDestroyHandler#destroy` doesn't use any `unIndex` method.

karreiro replied:
Fixed (The `unIndex` method were created and called in the pertinent places).

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/RecordEngine.java:
```diff
@@ -16,11 +16,30 @@
+    /**
+     * Destroys a record
+     * @param record
+     * @return returning an array of all affected records by the update operation.
```
by the destroy operation

karreiro replied:
Fixed.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/DataTypeHandlerTest.java:
```diff
@@ -0,0 +1,294 @@
+        final Optional<DataType> closestTopLevelDataType = handler.getClosestTopLevelDataType(dataType1);
+        final DataType actualDataType = closestTopLevelDataType.orElseThrow(UnsupportedOperationException::new);
+
+        assertEquals(dataType4, actualDataType);
```
Would you mind to check also if closest top level data type for dataType3 is dataType6?

karreiro replied:
Fixed.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java:
```diff
@@ -0,0 +1,84 @@
+                return Optional.of(parent);
+            } else {
+                for (final ItemDefinition itemDefinition : itemDefinitions()) {
+                    if (itemDefinition.getName().getValue().equals(parent.getTypeRef().getLocalPart())) {
```
Please use `Objects.equals`

karreiro replied:
Fixed.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtils.java:
```diff
@@ -50,4 +52,21 @@ public ItemDefinitionUtils(final DMNGraphUtils dmnGraphUtils) {
+                .orElseGet(() -> new QName(namespace, localPart, typeRefPrefix));
+    }
+
+    Optional<String> getPrefixForNamespaceURI(final String namespace) {
```
Technically this isn't 100% correct; but I think based on what we expose in our editor it is good enough.

```ItemDefinition``` extends ```DMNModelInstrumentedBase``` and hence can have its own Name Spaces defined (e.g. FEEL URI ```http://www.omg.org/spec/FEEL/20140401``` _could_ map to ```cheese``` in the ```ItemDefinition``` and hence we'd need the ```QName``` to be ```cheese:string```). *However* we do not support editing the Name Spaces for an ```ItemDefinition``` and hence always using ```Definitions``` is good enough :-)

karreiro replied:
Ok.. thank you, @manstis :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java:
```diff
@@ -0,0 +1,278 @@
+/**
+ * Manages all operations that change a Data Type.
+ * <p>
+ * A Data Type is a plan entity that must simple as simple as possible.
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java:
```diff
@@ -0,0 +1,278 @@
+        return itemDefinition.orElseThrow(() -> new UnsupportedOperationException("The data type must have an indexed ItemDefinition."));
+    }
+
+    private List<DataType> itemDefinitionSubDataTypes() {
```
This method creates sub data types from the `this.itemDefinition` attribute.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeManager.java:
```diff
@@ -0,0 +1,278 @@
+
+    ItemDefinition getItemDefinitionWithItemComponent(final ItemDefinition itemDefinition) {
+        if (itemDefinition.getTypeRef() != null) {
+            final Optional<ItemDefinition> definition = findByName(itemDefinition.getTypeRef().getLocalPart());
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/common/ItemDefinitionUtilsTest.java:
```diff
@@ -65,6 +70,82 @@ public void testFindByName() {
+        final String expectedPrefix = "feel";
+        final String expectedNamespace = "";
+        final String namespaceURI = FEEL.getUri();
+        final QName qName = new QName(namespaceURI, expectedLocalPart);
```
@jomarko here we're instantiating the parameter, the _real `QName`_ is instantiated inside of the `normaliseTypeRef` method, and that's we're asserting here :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/common/ElementHelper.java:
```diff
@@ -24,4 +24,8 @@ public static void insertAfter(final Node newNode,
     }
+
+    public static void remove(final Node element) {
+        element.parentNode.removeChild(element);
```
Thanks :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/persistence/handlers/ItemDefinitionDestroyHandler.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    List<ItemDefinition> itemDefinitions() {
+        return dmnGraphUtils.getDefinitions().getItemDefinition();
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItemTest.java:
```diff
@@ -232,28 +240,110 @@ public void testCloseEditMode() {
+    }
+
+    @Test
+    public void testRemoveTopLevelDataTypes() {
```
Fixed.

jomarko commented:
Could someone please add description into [DROOLS-2950](https://issues.jboss.org/browse/DROOLS-2950), otherwise I am not able to verify that this PR fixes that jira. Other manual checks are done from my site and found issues are reported.
